pkg/login/ext_user.go

```diff
@@ -0,0 +1,175 @@
+package login
+
+import (
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/log"
+	m "github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/quota"
+)
+
+func init() {
+	bus.AddHandler("auth", UpsertUser)
+}
+
+func UpsertUser(cmd *m.UpsertUserCommand) error {
+	extUser := cmd.ExternalUser
+
+	userQuery := &m.GetUserByAuthInfoQuery{
+		AuthModule: extUser.AuthModule,
+		AuthId:     extUser.AuthId,
+		UserId:     extUser.UserId,
+		Email:      extUser.Email,
+		Login:      extUser.Login,
+	}
+	err := bus.Dispatch(userQuery)
+	if err != nil {
+		if err != m.ErrUserNotFound {
+			return err
+		}
+
+		if !cmd.SignupAllowed {
+			log.Warn(fmt.Sprintf("Not allowing %s login, user not found in internal user database and allow signup = false", extUser.AuthModule))
+			return ErrInvalidCredentials
+		}
+
+		limitReached, err := quota.QuotaReached(cmd.ReqContext, "user")
+		if err != nil {
+			log.Warn("Error getting user quota", "err", err)
+			return ErrGettingUserQuota
+		}
+		if limitReached {
+			return ErrUsersQuotaReached
+		}
+
+		cmd.Result, err = createUser(extUser)
+		if err != nil {
+			return err
+		}
+
+		if extUser.AuthModule != "" && extUser.AuthId != "" {
+			cmd2 := &m.SetAuthInfoCommand{
+				UserId:     cmd.Result.Id,
+				AuthModule: extUser.AuthModule,
+				AuthId:     extUser.AuthId,
+			}
+			if err := bus.Dispatch(cmd2); err != nil {
+				return err
+			}
+		}
+	} else {
+		cmd.Result = userQuery.Result
+
+		// sync user info
+		err = updateUser(cmd.Result, extUser)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = syncOrgRoles(cmd.Result, extUser)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createUser(extUser *m.ExternalUserInfo) (*m.User, error) {
+	cmd := &m.CreateUserCommand{
+		Login: extUser.Login,
+		Email: extUser.Email,
+		Name:  extUser.Name,
+	}
+	if err := bus.Dispatch(cmd); err != nil {
+		return nil, err
+	}
+
+	return &cmd.Result, nil
+}
+
+func updateUser(user *m.User, extUser *m.ExternalUserInfo) error {
+	// sync user info
+	updateCmd := &m.UpdateUserCommand{
+		UserId: user.Id,
+	}
+	needsUpdate := false
+
+	if extUser.Login != "" && extUser.Login != user.Login {
+		updateCmd.Login = extUser.Login
+		needsUpdate = true
+	}
+	if extUser.Email != "" && extUser.Email != user.Email {
+		updateCmd.Email = extUser.Email
+		needsUpdate = true
+	}
+	if extUser.Name != "" && extUser.Name != user.Name {
+		updateCmd.Name = extUser.Name
+		needsUpdate = true
+	}
+
+	if needsUpdate {
+		log.Debug("Syncing user info", "id", user.Id, "update", updateCmd)
+		err := bus.Dispatch(updateCmd)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func syncOrgRoles(user *m.User, extUser *m.ExternalUserInfo) error {
```

Comment:
I don't see any tests for the syncing of org roles. In the old version, there were tests in [ldap_test.go](https://github.com/grafana/grafana/blob/master/pkg/login/ldap_test.go#L88-L164). Is this function covered in other tests now?